app/controllers/account_controller.rb

```diff
@@ -259,32 +260,15 @@ def activate_through_ldap(user)
   # to change the password.
   # When making changes here, also check MyController.change_password
   def change_password
-    return render_404 if OpenProject::Configuration.disable_password_login?
-
-    @user = User.find_by_login(params[:username])
-    @username = @user.login
-
-    # A JavaScript hides the force_password_change field for external
-    # auth sources in the admin UI, so this shouldn't normally happen.
-    return if redirect_if_password_change_not_allowed(@user)
-
-    if @user.check_password?(params[:password])
-      @user.password = params[:new_password]
-      @user.password_confirmation = params[:new_password_confirmation]
-      @user.force_password_change = false
-      if @user.save
-
-        result = password_authentication(params[:username], params[:new_password])
-        # password_authentication resets session including flash notices,
-        # so set afterwards.
-        flash[:notice] = l(:notice_account_password_updated)
-        return result
-      end
-    else
-      invalid_credentials
+    # Retrieve user_id from session
+    @user = User.find(flash[:_password_change_user_id])
+
+    change_password_flow(user: @user, params: params, show_user_name: true) do
+      password_authentication(@user.login, params[:new_password])
     end
-    # Render the username to hint to a user in case of a forced password change
-    render 'my/password', locals: { show_user_name: @user.force_password_change_was }
+  rescue ActiveRecord::RecordNotFound
+    Rails.logger.error "Failed to find user for change_password request: #{flash[:_password_change_user_id]}"
+    render_404
```

Comment:
Why just a plain 404 here instead of redirecting the user back and showing a flash message?

Reply:
Because it should never happen in a regular flow that posting to the form without the session value being present.